### PR TITLE
[luci-interpreter] Add Maximum kernel

### PR DESCRIPTION
**Pull request by underflow101**
Parent Issue: #3458
Fired Issue: #4135 

This commit adds `Maximum` kernel.

This commit have been tested on my PC,
and worked well by passing `./nncc build` & `./nncc test`.

<pre><code>
void Maximum(const RuntimeShape& input1_shape, const T* input1_data,
             const T* input2_data, const RuntimeShape& output_shape,
             T* output_data)</code></pre>
`reference_op` above didn't work well giving the error below:
<details><summary><b>Error when using reference_op</b></summary>
<pre><code>
ONE/compiler/luci-interpreter/src/kernels/Maximum.test.cpp:51: Failure
Value of: extractTensorData<float>(output_tensor)
Expected: has 6 elements where
element #0 is approximately 1 (absolute error <= 9.9999997e-06),
element #1 is approximately 0 (absolute error <= 9.9999997e-06),
element #2 is approximately 1 (absolute error <= 9.9999997e-06),
element #3 is approximately 12 (absolute error <= 9.9999997e-06),
element #4 is approximately -2 (absolute error <= 9.9999997e-06),
element #5 is approximately -1.4299999 (absolute error <= 9.9999997e-06)
  Actual: { 1, 0, -1, 11, -1, -1 }, whose element #2 doesn't match, which is -2 from 1
</code></pre></details>

So I decided to use `std::max()` function to write this code.

Please review this PR, @seanshpark @jinevening @s-barannikov  .
I'll be happy to get feedback and change to make improvement.

Thank you.

(Derived from SOS Mine Project)

Signed-off-by: underflow101 <ikarus125@gmail.com>

### CODE REVIEW

**Review by s-barannikov — COMMENTED** [review]
AFAIK, Maximum should support broadcasting. Please refer to Add or Mul for an example.
Please also add a test to KernelBuilder.test.cpp.

**Comment by underflow101** [comment]
> AFAIK, Maximum should support broadcasting. Please refer to Add or Mul for an example.
> Please also add a test to KernelBuilder.test.cpp.

Thank you for your advice, @s-barannikov ,
I'll do that.

File: compiler/luci-interpreter/src/kernels/Maximum.test.cpp
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Maximum.test.cpp:32]: We usually use snake_case for variables. Also, why don't just `return (max - min) / 255.0`?

Comment (underflow101) [reply compiler/luci-interpreter/src/kernels/Maximum.test.cpp:32]: Actually, thank you for pointing that out.
`GetTolerance()` is never used in Maximum.
I will remove this.

Thank you! :)

**Comment by underflow101** [comment]
@s-barannikov @jinevening ,

First of all, I'm so sorry for too late fix of my PR.
I am truly sorry for leaving my PR for more than a month.
It will not happen again.

I have added broadcast support, but `MaximumMinimumBroadcastSlow()` needs `Op` parameter,
So I have put `tflite::ops::builtin::maximum_minimum::MaximumOp()` for `Op` parameter,

But it returns the error below:

<details><summary><b>Error Message</b></summary>
<pre><code>In file included from /home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/maximum_minimum.cc:15,
                 from /home/bear/data/opensource/ONE/compiler/luci-interpreter/src/kernels/Maximum.cpp:22:
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h: In instantiation of ‘void tflite::reference_ops::MaximumMinimumBroadcastSlow(const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, T*, Op) [with T = float; Op = tflite::ops::builtin::maximum_minimum::MaximumOp; int N = 5]’:
/home/bear/data/opensource/ONE/compiler/luci-interpreter/src/kernels/Maximum.cpp:70:149:   required from here
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h:37:26: error: no match for call to ‘(tflite::ops::builtin::maximum_minimum::MaximumOp) (const float&, const float&)’
   37 |       output_data[i] = op(input1_data[i], input2_data[i]);
      |                        ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h:54:13: error: no match for call to ‘(tflite::ops::builtin::maximum_minimum::MaximumOp) (const float&, const float&)’
   54 |           op(input1_data[SubscriptToIndex(desc1, indexes)],
      |           ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   55 |              input2_data[SubscriptToIndex(desc2, indexes)]);
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h: In instantiation of ‘void tflite::reference_ops::MaximumMinimumBroadcastSlow(const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, T*, Op) [with T = unsigned char; Op = tflite::ops::builtin::maximum_minimum::MaximumOp; int N = 5]’:
/home/bear/data/opensource/ONE/compiler/luci-interpreter/src/kernels/Maximum.cpp:92:153:   required from here
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h:37:26: error: no match for call to ‘(tflite::ops::builtin::maximum_minimum::MaximumOp) (const unsigned char&, const unsigned char&)’
   37 |       output_data[i] = op(input1_data[i], input2_data[i]);
      |                        ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h:54:13: error: no match for call to ‘(tflite::ops::builtin::maximum_minimum::MaximumOp) (const unsigned char&, const unsigned char&)’
   54 |           op(input1_data[SubscriptToIndex(desc1, indexes)],
      |           ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   55 |              input2_data[SubscriptToIndex(desc2, indexes)]);
      |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/reference/maximum_minimum.h:18,
                 from /home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/maximum_minimum.cc:15,
                 from /home/bear/data/opensource/ONE/compiler/luci-interpreter/src/kernels/Maximum.cpp:22:
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/common.h:832:13: error: ‘void tflite::NDOpsHelper(const tflite::NdArrayDesc<N>&, const Calc&) [with int N = 5; Calc = tflite::reference_ops::MaximumMinimumBroadcastSlow(const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, T*, Op) [with T = float; Op = tflite::ops::builtin::maximum_minimum::MaximumOp; int N = 5]::<lambda(int*)>]’, declared using local type ‘const tflite::reference_ops::MaximumMinimumBroadcastSlow(const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, T*, Op) [with T = float; Op = tflite::ops::builtin::maximum_minimum::MaximumOp; int N = 5]::<lambda(int*)>’, is used but never defined [-fpermissive]
  832 | inline void NDOpsHelper(const NdArrayDesc<N>& output, const Calc& calc) {
      |             ^~~~~~~~~~~
/home/bear/data/opensource/ONE/externals/TENSORFLOW-2.3.0/tensorflow/lite/kernels/internal/common.h:832:13: error: ‘void tflite::NDOpsHelper(const tflite::NdArrayDesc<N>&, const Calc&) [with int N = 5; Calc = tflite::reference_ops::MaximumMinimumBroadcastSlow(const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, T*, Op) [with T = unsigned char; Op = tflite::ops::builtin::maximum_minimum::MaximumOp; int N = 5]::<lambda(int*)>]’, declared using local type ‘const tflite::reference_ops::MaximumMinimumBroadcastSlow(const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, const T*, const tflite::RuntimeShape&, T*, Op) [with T = unsigned char; Op = tflite::ops::builtin::maximum_minimum::MaximumOp; int N = 5]::<lambda(int*)>’, is used but never defined [-fpermissive]
make[2]: *** [compiler/luci-interpreter/src/kernels/CMakeFiles/luci_interpreter_kernels.dir/build.make:271: compiler/luci-interpreter/src/kernels/CMakeFiles/luci_interpreter_kernels.dir/Maximum.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:15737: compiler/luci-interpreter/src/kernels/CMakeFiles/luci_interpreter_kernels.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
</code></pre></details>

I have double-checked `externals/tensorflow-2.3.0/tensorflow/lite/kernels/maximum_minimum.cc`, but could not find any other form of `op` parameter than `MaximumOp`.
Could you please give me an advice?

**Comment by s-barannikov** [comment]
Well, `MaximumMinimumBroadcastSlow` expects callable as an `op` argument, and you are passing an instance of a class, which is not callable. You probably wanted to pass `MaximumOp::template op<float>` instead.
In any case, it is not a good idea to #include *.cc file, you will get linker errors.
The solution is to write your own version of `MaximumOp` class and use it like the above, or a simple lambda, like
```C++
[](float x, float y) { return std::max(x, y); }
```

Also, I would suggest to wait until #4567 is merged and use it instead of `maximum_minimum.h`. You can find example usage in #4568.

**Comment by underflow101** [comment]
> Well, `MaximumMinimumBroadcastSlow` expects callable as an `op` argument, and you are passing an instance of a class, which is not callable. You probably wanted to pass `MaximumOp::template op<float>` instead.
> In any case, it is not a good idea to #include *.cc file, you will get linker errors.
> The solution is to write your own version of `MaximumOp` class and use it like the above, or a simple lambda, like
> 
> ```c++
> [](float x, float y) { return std::max(x, y); }
> ```
> 
> Also, I would suggest to wait until #4567 is merged and use it instead of `maximum_minimum.h`. You can find example usage in #4568.

Thank you so much for your advice, @s-barannikov ,
I was thinking inside the box that I couldn't write my own version of template (at least I thought I wasn't allowed to do so)

I will work on it and wait until #4567 is merged first, and then rebase my branch before I commit my editted version.
Thank you so much again for your advice! :)

**Comment by underflow101** [comment]
@s-barannikov , @jinevening ,
I have referred to #4568 for this commit.
Thank you for your advice and examples, @s-barannikov :)

Please take a look for this commit.

Thank you so much.

File: compiler/luci-interpreter/src/kernels/Maximum.cpp
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Maximum.cpp:70]: I think you can just in-line these two methods into `execute`, as they are very small.

**Review by s-barannikov — COMMENTED** [review]
Please add a test to `compiler/luci-interpreter/src/loader/KernelBuilder.test.cpp`. Otherwise looks nice, thank you.

File: compiler/luci-interpreter/src/kernels/Maximum.cpp
Comment (underflow101) [reply compiler/luci-interpreter/src/kernels/Maximum.cpp:70]: I think in-lining these code could harm the source uniformity (since other kernel sources keep this format)
Don't you think it's better to keep this format?

I will follow your decision.
Thank you.

**Comment by underflow101** [comment]
> Please add a test to `compiler/luci-interpreter/src/loader/KernelBuilder.test.cpp`. Otherwise looks nice, thank you.

@s-barannikov ,
By adding `Maximum` to `KernelBuilder.cpp`, `KernelBuilder.h` & `KernelBuild.test.cpp`, it causes build error due to pre-existing `Maximum_U8_000` recipe.

I will fix `Maximum_U8_000` after this PR is merged, and then re-commit KernelBuilder sources.

Thank you.

File: compiler/luci-interpreter/src/kernels/Maximum.cpp
Comment (s-barannikov) [reply compiler/luci-interpreter/src/kernels/Maximum.cpp:70]: > other kernel sources keep this format

Most of them are, but that's not necessary and certainly not a rule (see DepthToSpace, Squeeze, Unpack and some others). We usually split `execute` method into several `eval*` methods when keeping everything in `execute` method would make it too complex to read / understand. This one looks very simple -- just one function call; I would not make separate methods just for a function call :)

NB
Uniformity is important, but it also has downside -- it is hard to keep it. Once you changed one file, you will have to look through all others and check if they need to be changed too.
Things get worse when you work in parallel with other developers: you might have finally changed all the files and created a PR, but at the same time there could be parallel PR which did not take your changes into account. After you both merged, there will be inconsistency again.

**Comment by s-barannikov** [comment]
> By adding `Maximum` to `KernelBuilder.cpp`, `KernelBuilder.h` & `KernelBuild.test.cpp`, it causes build error due to pre-existing `Maximum_U8_000` recipe.
> 
> I will fix `Maximum_U8_000` after this PR is merged, and then re-commit KernelBuilder sources.

Not sure I got the problem... I think recipies have nothing to do with your PR. Could you show the error log?

File: compiler/luci-interpreter/src/kernels/Maximum.test.cpp
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Maximum.test.cpp:30]: Looks unused :)

**Comment by underflow101** [comment]
> > By adding `Maximum` to `KernelBuilder.cpp`, `KernelBuilder.h` & `KernelBuild.test.cpp`, it causes build error due to pre-existing `Maximum_U8_000` recipe.
> > I will fix `Maximum_U8_000` after this PR is merged, and then re-commit KernelBuilder sources.
> 
> Not sure I got the problem... I think recipies have nothing to do with your PR. Could you show the error log?

I'm Sorry for my misunderstanding. That recipe only existed on my local PC.

File: compiler/luci-interpreter/src/kernels/Maximum.test.cpp
Comment (underflow101) [reply compiler/luci-interpreter/src/kernels/Maximum.test.cpp:30]: Thank you! I will remove this :)

File: compiler/luci-interpreter/src/kernels/Maximum.cpp
Comment (underflow101) [reply compiler/luci-interpreter/src/kernels/Maximum.cpp:70]: > NB
> Uniformity is important, but it also has downside -- it is hard to keep it. Once you changed one file, you will have to look through all others and check if they need to be changed too.
> Things get worse when you work in parallel with other developers: you might have finally changed all the files and created a PR, but at the same time there could be parallel PR which did not take your changes into account. After you both merged, there will be inconsistency again.

Thank you for your sincere and kind advice, @s-barannikov ,
It is meaningful to me to have conversations like this with you :)

I will put that in my mind when I develop with other people.
Thank you so much! :)

**Review by s-barannikov — APPROVED** [review]
LGTM

**Review by jinevening — APPROVED** [review]
LGTM

Sorry for the late review. I was on vacation yesterday.

**Comment by llFreetimell** [comment]
@seanshpark Merge plz :)

**Pull request merged**
